xarray/core/dataset.py

```diff
@@ -1210,10 +1210,20 @@ def _construct_dataarray(self, name: Hashable) -> DataArray:
         needed_dims = set(variable.dims)
 
         coords: dict[Hashable, Variable] = {}
+        temp_indexes = self.xindexes
         # preserve ordering
         for k in self._variables:
-            if k in self._coord_names and set(self._variables[k].dims) <= needed_dims:
-                coords[k] = self._variables[k]
+            if k in self._coord_names:
```

Comment:
I'm surprised this is so much cleaner than the dataarray case (which has a nested loop). Is the approach here applicable to dataarray?